Symfony autocomplete

ecairol opened the pull request:
WIP

ecairol committed:
Dummy testing command to try out Symfony auto-complete

ecairol committed:
First proposal for autocomplete using three commands as example

ecairol commented:
@ahegyes how do you feel about this. When the user writes a command and hits `TAB`, besides displaying the available options and arguments, to also offer the user a full written example with placeholders?

<img width="1086" alt="Screenshot 2024-02-05 at 10 55 14" src="https://github.com/Automattic/team51-cli/assets/1821060/3407e63a-ff91-4fca-b345-8f266193154e">

--
At least on my Terminal the options are not selectable, and I think it could be useful to have a reference of the value types that can be used for each option, but I'm also having second thoughts about it.

ahegyes commented:
> how do you feel about this

I'm in favor of not doing that simply because I think it kinda clutters the terminal and I haven't seen it done anywhere else.

But I can see some value in it. However, is the example auto-generated or will that be another thing to remember to update if the options change? 🤔

ecairol commented:
@ahegyes Yeah that's ok, it was an idea I had and wanted to hear opinions, but I wasn't 100% sold on it either. I haven't found a way to make it auto-generated so that's a con, it would need to be manually updated 👎

ecairol committed:
Added Autocomplete trait with generic complete function and use it on all commands

ecairol commented:
@ahegyes some recent updates:
 - I've implemented a Trait called Autocomplete, with a generic `complete()` function, and now all commands are using it. It works like a charm
 - Removed the "example" part from the auto-complete
 - Now the `--help` command skips the 1Password part and returns results much faster
 - There was a conflict with Symphony autocomplete and the `-c` command, so I had to rename the `--contractor` option to `-con`

Would you mind testing when you have a chance?

ecairol committed:
Return int value on execute() functions of all commands.

Also added int return type and changed.

ecairol commented:
@fmfernandes since you're already familiar with this problem and its solution, would you mind taking a look at commit [eae6cb4](https://github.com/Automattic/team51-cli/pull/217/commits/eae6cb45b447f8abb262db85a8d4b70e6c9ef891) and let me know if you see anything suspicious or have any suggestion or question?

fmfernandes commented on src/commands/pressable-call-api.php:
```diff
@@ -47,20 +47,22 @@ private function get_question_helper(): QuestionHelper {
 		if ( ! in_array( $input->getOption( 'format' ), array( 'text', 'json' ), true ) ) {
 			$this->output->writeln( '<error>Invalid output format</error>' );
-			return;
+			return Command::FAILURE;
```
According to Symfony [docs](https://symfony.com/doc/current/console.html#creating-a-command), we might want to return `Command::INVALID` since we're dealing with an unexpected input from the user.

fmfernandes commented on src/commands/wpcom-get-stickers.php:
```diff
@@ -58,14 +58,14 @@ protected function initialize( InputInterface $input, OutputInterface $output ):
 		if ( empty( $stickers ) ) {
 			$output->writeln( '<comment>Site has no stickers associated.<comment>' );
-			exit;
+			return Command::FAILURE;
```
Similarly, I'd return `Command::SUCCESS` here since not having an sticker is not an issue.

fmfernandes reviewed (commented):
PR looks good. Left a couple comments regarding some return types I'd change to match the command's output.

ecairol committed:
Updated a couple of returned responses from failure to success or invalid

ecairol commented:
Thanks for those observations @fmfernandes, I've updated both cases here: [c5978c9](https://github.com/Automattic/team51-cli/pull/217/commits/c5978c9d9a8024e2a6253ab0a3d526ecc74e5912)
